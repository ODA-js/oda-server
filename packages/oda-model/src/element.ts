import { IValidate, IValidator, IValidationResultInput } from './validation';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';

export const Internal = Symbol.for('internal');
export const MetaData = Symbol.for('metadata');

export interface ElementMetaInfo {}

export interface ElementInternal<T extends ElementMetaInfo> {
  [MetaData]: T;
}

export interface ElementInput<T extends ElementMetaInfo> {
  metadata?: Partial<T>;
}

export interface ElementOutput<T extends ElementMetaInfo> {
  metadata: T;
}

export interface IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: any): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): any;
}

/**
 * updatable items
 */
export interface IUpdatable<
  T extends ElementMetaInfo,
  I extends ElementInput<T>,
  O extends ElementOutput<T>
> extends IUpdatableBase {
  /**
   * update item with data
   * @param payload the update payload
   */
  updateWith(payload: Nullable<I>): void;
  /**
   * return copy of object item that is suitable for creating new one
   */
  toObject(): O;
}

/**
 * Meta interface for updating items with metadata
 */
export interface IMeta<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  O extends ElementOutput<M>
> extends IUpdatable<M, I, O>, IValidate {
  /**
   * the kind of current item
   */
  readonly modelType: MetaModelType;
  /**
   * meta information that is outside of model notations and can be customized as well
   */
  readonly metadata: Readonly<M>;
}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Element<
  M extends ElementMetaInfo,
  I extends ElementInput<M>,
  S extends ElementInternal<M>,
  O extends ElementOutput<M>
> implements IMeta<M, I, O> {
  public get modelType(): MetaModelType {
    return 'element';
  }
  public [Internal]: S;
  /**
   * metadata is not immutable
   */
  public get metadata(): M {
    return this[Internal][MetaData];
  }

  public validate(validator: IValidator): IValidationResultInput[] {
    return validator.check(this);
  }

  constructor(init: ElementInput<M>) {
    this[Internal] = {} as S;
    this.updateWith(merge({}, defaultInput, init));
  }
  /**
   * patches metadata for entity
   * @param input metadata patch object
   */
  public updateWith(input: Nullable<ElementInput<M>>) {
    assignValue<Element<M, I, S, O>, ElementInput<M>, ElementInternal<M>>({
      src: this,
      input,
      field: 'metadata',
      effect: (_, value) =>
        (this[Internal][MetaData] = merge({}, this[Internal][MetaData], value)),
      setDefault: _ =>
        (this[Internal][MetaData] = merge({}, defaultMetaInfo) as M),
    });
  }
  /**
   * make item usable within another constructor
   * it is suitable for cloning items
   */
  public toObject(): O {
    return merge({}, { metadata: this.metadata } as O);
  }
}
