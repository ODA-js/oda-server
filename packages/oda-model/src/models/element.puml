@startuml

interface IMeta
interface IUpdatable
interface IValidate

IUpdatable <|-- IMeta
IValidate <|-- IMeta

class Element

IMeta <|-- Element


@enduml
