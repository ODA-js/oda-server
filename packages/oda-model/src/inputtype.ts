import { ElementMetaInfo, Internal, MetaData } from './element';
import {
  IModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import {
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  HashToArray,
  NamedArray,
  MapToArray,
  ArrayToMap,
} from './types';

import { merge } from 'lodash';
import * as inflected from 'inflected';
import { InputFieldInput, InputField, IInputField } from './inputfield';
import { EntityFieldInput } from './entityfield';
import { UIView } from './entitybase';

export interface IInputType
  extends IModelBase<InputTypeMetaInfo, InputTypeInput, InputTypeOutput> {
  readonly plural: string;
  readonly global: boolean;
  readonly titlePlural: string;
  readonly fields: Map<string, IInputField>;
}

export interface InputTypeMetaInfo extends ElementMetaInfo {
  titlePlural: string;
  global: boolean;
  name: {
    plural: string;
    singular: string;
  };
  UI: UIView;
}

export interface InputTypeInternal extends ModelBaseInternal {
  fields: Map<string, IInputField>;
}

export interface InputTypeInput extends ModelBaseInput<InputTypeMetaInfo> {
  plural?: string;
  titlePlural?: string;
  fields?: AsHash<InputFieldInput> | NamedArray<InputFieldInput>;
}

export interface InputTypeOutput extends ModelBaseOutput<InputTypeMetaInfo> {
  fields: NamedArray<InputFieldInput>;
}

export const InputTypeDefaultMetaInfo = {
  name: {},
  UI: {
    listName: [],
    quickSearch: [],
    hidden: [],
    edit: [],
    show: [],
    list: [],
    embedded: [],
  },
};
export const InputTypeDefaultInput = {
  metadata: InputTypeDefaultMetaInfo,
  exact: false,
};

export class InputType
  extends ModelBase<
    InputTypeMetaInfo,
    InputTypeInput,
    InputTypeInternal,
    InputTypeOutput
  >
  implements IInputType {
  public get modelType(): MetaModelType {
    return 'input-type';
  }

  constructor(init: InputTypeInput) {
    super(merge({}, InputTypeDefaultInput, init));
  }

  get plural(): string {
    return this.metadata.name.plural;
  }

  get titlePlural(): string {
    return this.metadata.titlePlural;
  }

  get global(): boolean {
    return this.metadata.global;
  }

  get fields(): Map<string, IInputField> {
    return this[Internal].fields;
  }

  public updateWith(input: Nullable<InputTypeInput>) {
    super.updateWith(input);

    assignValue<InputTypeInternal, InputTypeInput, string>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => {
        src.name = value.trim();
        this.metadata.name.singular = src.name;
      },
      required: true,
    });

    assignValue<InputTypeMetaInfo, InputTypeInput, boolean>({
      src: this[MetaData],
      input,
      field: 'name',
      effect: (src, value) => (src.global = value),
      setDefault: src => (src.global = false),
    });

    assignValue<InputTypeMetaInfo, InputTypeInput, string>({
      src: this.metadata,
      input,
      field: 'name.plural',
      inputField: 'plural',
      effect: (src, value) => {
        src.name.plural = value.trim();
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
      setDefault: src => {
        src.name.plural = inflected.pluralize(this.name);
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
    });

    assignValue<InputTypeMetaInfo, InputTypeInput, string>({
      src: this.metadata,
      input,
      field: 'titlePlural',
      effect: (src, value) => (src.titlePlural = value.trim()),
      setDefault: src => (src.titlePlural = this.plural),
    });

    assignValue<
      InputTypeInternal,
      InputTypeInput,
      AsHash<InputFieldInput> | NamedArray<InputFieldInput>
    >({
      src: this[Internal],
      input,
      field: 'fields',
      effect: (src, value) => {
        const fields = ArrayToMap(
          Array.isArray(value) ? value : HashToArray(value),
          (fld, order) => {
            let field = new InputField(
              merge({}, fld, {
                order,
                entity: this.name,
              } as Partial<EntityFieldInput>),
            );
            return field;
          },
          (obj: any, src: any) => obj.mergeWith(src.toObject()),
        );

        src.fields = fields;
      },
      setDefault: src => (src.fields = new Map()),
    });
  }

  public toObject(): InputTypeOutput {
    return merge({}, super.toObject(), {
      fields: MapToArray(this.fields, (name, value) => ({
        ...value.toObject(),
        name,
      })),
    } as Partial<InputTypeOutput>);
  }

  public mergeWith(payload: Nullable<InputTypeInput>) {
    super.mergeWith(payload);
  }
}
