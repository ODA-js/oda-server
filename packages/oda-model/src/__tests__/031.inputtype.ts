import 'jest';
import { MetaData } from '../element';
import { InputType, InputTypeDefaultMetaInfo } from '../inputtype';

describe('InputType', () => {
  it('default', () => {
    const res = new InputType({ name: 'A' });
    expect(res[MetaData]).toMatchObject(InputTypeDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('default has no Singular', () => {
    const res = new InputType({ name: 'Species' });
    expect(res.plural).toBe('AllSpecies');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('should Capitalize name', () => {
    const res = new InputType({ name: 'entity' });
    expect(res.name).toBe('entity');
  });
  it('take TitlePlural from Plural name', () => {
    const res = new InputType({ name: 'Child', plural: 'Children  ' });
    expect(res.plural).toBe('Children');
    expect(res.titlePlural).toBe('Children');
  });
  it('should take Plural Title plural', () => {
    const res = new InputType({
      name: 'Species',
      plural: 'Species',
      titlePlural: 'All Species',
    });
    expect(res.titlePlural).toBe('All Species');
    expect(res.plural).toBe('AllSpecies');
  });

  it('creates specified field with specified types', () => {
    const res = new InputType({
      name: 'A',
      fields: [
        {
          name: 'A',
        },
        {
          name: 'B',
          type: { type: 'enum', name: 'BBB' },
        },
      ],
    });
    expect(res.fields.size).toBe(2);
    expect(res.fields.get('A')).not.toBeUndefined();
    const a = res.fields.get('a');
    if (a) {
      expect(a.modelType).toBe('input-simple-field');
    }
    expect(res.fields.get('A')).not.toBeUndefined();
    const b = res.fields.get('B');
    if (b) {
      expect(b.modelType).toBe('input-enum-field');
    }
    expect(res).toMatchSnapshot('object with fields');
    expect(res.toObject()).toMatchSnapshot('toObject with fields');
  });

  it('should pass entity name to fields', () => {
    const res = new InputType({
      name: 'A',
      fields: [{ name: 'a' }],
    });
    const a = res.fields.get('a');
    if (a) {
      expect(a.metadata.entity).toBe('A');
      expect(a.metadata.order).toBe(0);
    }
  });

  it('should deduplicates fields', () => {
    const res = new InputType({
      name: 'A',
      fields: [
        { name: 'a' },
        { name: 'a', description: 'field a description' },
      ],
    });
    expect(res.fields.size).toBe(1);
  });
});
