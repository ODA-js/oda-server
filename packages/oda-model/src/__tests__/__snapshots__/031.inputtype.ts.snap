// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputType creates specified field with specified types: object with fields 1`] = `
InputType {
  Symbol(internal): Object {
    "fields": Map {
      "A" => InputField {
        Symbol(internal): Object {
          "list": false,
          "name": "A",
          "type": "string",
        },
        Symbol(metadata): Object {
          "entity": "A",
          "order": 0,
        },
      },
      "B" => InputField {
        Symbol(internal): Object {
          "list": false,
          "name": "B",
          "type": Object {
            "multiplicity": "one",
            "name": "BBB",
            "type": "enum",
          },
        },
        Symbol(metadata): Object {
          "entity": "A",
          "order": 1,
        },
      },
    },
    "name": "A",
  },
  Symbol(metadata): Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
}
`;

exports[`InputType creates specified field with specified types: toObject with fields 1`] = `
Object {
  "description": undefined,
  "fields": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "list": false,
      "metadata": Object {
        "entity": "A",
        "order": 0,
      },
      "name": "A",
      "title": undefined,
      "type": "string",
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "list": false,
      "metadata": Object {
        "entity": "A",
        "order": 1,
      },
      "name": "B",
      "title": undefined,
      "type": Object {
        "multiplicity": "one",
        "name": "BBB",
        "type": "enum",
      },
    },
  ],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
  "name": "A",
  "title": undefined,
}
`;

exports[`InputType default has no Singular: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "name": Object {
      "plural": "AllSpecies",
      "singular": "Species",
    },
    "titlePlural": "AllSpecies",
  },
  "name": "Species",
  "title": undefined,
}
`;

exports[`InputType default: toObject 1`] = `
Object {
  "description": undefined,
  "fields": Array [],
  "metadata": Object {
    "UI": Object {
      "edit": Array [],
      "embedded": Array [],
      "hidden": Array [],
      "list": Array [],
      "listName": Array [],
      "quickSearch": Array [],
      "show": Array [],
    },
    "name": Object {
      "plural": "As",
      "singular": "A",
    },
    "titlePlural": "As",
  },
  "name": "A",
  "title": undefined,
}
`;
